File: docs/palindromes-in-peg/scaffold_chain.py
"""Persistent semiperiod prediction driven by the real DP unary answer.

The input walker copies C..C-h locally. A private tape then bounces between
its marked ends; a separate readonly verifier catches up with the already
matched radius. No coordinates, head identity tests, or numeric h are used.
The caller owns center shifts and must update the relative counters with
shift_one() for each real place moved by its center.
"""
from fpp_finite import LEFT
from scaffold_program import TapeView
from scaffold_search import alias
from scavm_structs import CounterView


class ChainView:
  def __init__(self, vm, previous, builder, center, walker, verifier, radius):
    self.vm, self.builder = vm, builder
    self.center, self.walker, self.verifier, self.radius = center, walker, verifier, radius
    self.period = TapeView(vm, previous, builder, "ch.p")
    self.h, self.lag, self.distance, self.boundary, self.last, self.margin, self.cycle = [
      CounterView(vm, previous, builder, "ch." + name)
      for name in ("h", "lag", "dist", "bound", "last", "margin", "cycle")]
    if previous is None:
      self.mode, self.direction, self.phase = "idle", 1, 0
      self.period_only = False
    else:
      lab = vm.label(previous)
      self.mode, self.direction, self.phase = lab["ch.mode"], lab["ch.dir"], lab["ch.phase"]
      self.period_only = lab["ch.only"]

  def start(self):
    self.period.reset()
    self.period.write("F:" + self.center.read())
    self.walker.copy_from(self.center)
    self.verifier.copy_from(self.center)
    self.h.reset()
    self.distance.reset()
    self.boundary.reset()
    self.last.reset()
    self.cycle.reset()
    self.period_only = False
    alias(self.lag, self.radius)
    alias(self.margin, self.radius)
    self.mode, self.direction, self.phase = "copy", 1, 0

  def prediction(self):
    if self.mode != "watch" or self.lag.sign() != 0:
      return None
    return self.period.read()[-1]

  def can_shift(self):
    ready = self.mode == "watch" and self.lag.sign() == 0 and self.phase == 4
    return ready and (self.cycle_end() if self.period_only else self.margin.sign() >= 0)

  def cycle_end(self):
    return (self.cycle.sign() > 0 and
            self.cycle.pos.below_of_top() is None)

  def check_pair(self, left):
    """Check the two-semiperiod continuation invariant against real input."""
    if self.period_only and self.mode == "watch" and self.lag.sign() == 0:
      if self.cycle.sign() <= 0:
        raise AssertionError("chain continuation exhausted without dispatch")
      same = left == self.prediction()
      if same == self.cycle_end():
        raise AssertionError("left head contradicts the two-semiperiod continuation")

  def begin_shift(self):
    self.period_only = True
    self.cycle.reset()

  def _consume(self):
    self.verifier.right()
    token = self.period.read()
    if self.verifier.read() != token[-1]:
      self.mode = "broken"
      return False
    self.distance.inc()
    if token.startswith(("F:", "T:")):
      alias(self.last, self.boundary)
      alias(self.boundary, self.distance)
      self.phase = min(4, self.phase + 1)
      self.direction = 1 if token.startswith("F:") else -1
    self.period.move(self.direction)
    return True

  def matched(self):
    """One new place has joined the matched interval (also on a chain shift)."""
    self.margin.inc()
    if self.period_only:
      self.cycle.dec()
    if self.mode == "watch" and self.lag.sign() == 0:
      self._consume()
    else:
      self.lag.inc()

  def shift_one(self):
    for counter in (self.distance, self.boundary, self.last, self.margin):
      counter.dec()
    self.cycle.inc()
    self.cycle.inc()

  def step(self, answer):
    if self.mode == "copy":
      if answer.read() == "1":
        answer.move(-1)
        self.h.inc()
        for _ in range(4):
          self.margin.dec()
        self.walker.left()
        symbol = self.walker.read()
        if symbol is None:
          raise ValueError("DP answer crosses the input origin")
        self.period.move(1)
        self.period.write(symbol)
      elif answer.read() == LEFT and self.h.sign() > 0:
        self.period.write("T:" + self.period.read())
        self.mode = "back"
      else:
        raise ValueError("positive unary DP answer required")
    elif self.mode == "back":
      if self.period.read().startswith("F:"):
        self.period.move(1)
        self.mode = "watch"
      else:
        self.period.move(-1)
    elif self.mode == "watch" and self.lag.sign() > 0:
      if self._consume():
        self.lag.dec()
    elif self.mode not in ("idle", "watch", "broken"):
      raise AssertionError(self.mode)

  def finalize(self):
    self.period.finalize()
    for counter in (self.h, self.lag, self.distance, self.boundary, self.last, self.margin, self.cycle):
      counter.finalize()
    self.builder.label.update({"ch.mode": self.mode, "ch.dir": self.direction,
                               "ch.phase": self.phase, "ch.only": self.period_only})

File: scala/pal/src/main/scala/pal/ScaffoldChain.scala
package pal

import Scavm.{Label, Node, VM}
import ScavmStructs.{Builder, CounterView}
import ScaffoldInput.ReadonlyHead
import ScaffoldProgram.TapeView
import ScaffoldSearch.alias
import FppFinite.LEFT

/** Persistent semiperiod prediction driven by the real DP unary answer.
  *
  * The input walker copies C..C-h locally. A private tape then bounces between
  * its marked ends; a separate readonly verifier catches up with the already
  * matched radius. No coordinates, head identity tests, or numeric h are used.
  * The caller owns center shifts and must update the relative counters with
  * shift_one() for each real place moved by its center.
  *
  * Python 原典: `scaffold_chain.py`（`SCA_GALIL.md` "Implemented transitions"）。
  * Python の `finalize` は `finish`。
  */
object ScaffoldChain {

  /** The chain state, stored in the label `ch.mode` as its Python string. */
  enum Mode(val label: String) {
    case Idle extends Mode("idle")
    case Copy extends Mode("copy")
    case Back extends Mode("back")
    case Watch extends Mode("watch")
    case Broken extends Mode("broken")
  }

  object Mode {
    def fromLabel(label: Label): Mode = {
      val s = label.asStr
      values.find(_.label == s).getOrElse(throw new IllegalArgumentException(s"unknown chain mode '$s'"))
    }
  }

  /** The period tape token marking the first copied place (the center). */
  val FRONT_MARK: String = "F:"

  /** The period tape token marking the last copied place (C-h). */
  val TAIL_MARK: String = "T:"

  /** The chain view: a private period tape and seven unary counters, over a
    * readonly center, a walker that copies the semiperiod and a verifier that
    * catches up with the matched radius (all owned by the caller).
    */
  final class ChainView[H <: ReadonlyHead[H]](
      val vm: VM,
      previous: Option[Node],
      val builder: Builder,
      val center: H,
      val walker: H,
      val verifier: H,
      val radius: CounterView
  ) {

    val period: TapeView = new TapeView(vm, previous, builder, "ch.p")

    /** The unary semiperiod copied from the DP answer. */
    val h: CounterView = new CounterView(vm, previous, builder, "ch.h")
    /** How far the verifier lags behind the matched radius. */
    val lag: CounterView = new CounterView(vm, previous, builder, "ch.lag")
    val distance: CounterView = new CounterView(vm, previous, builder, "ch.dist")
    val boundary: CounterView = new CounterView(vm, previous, builder, "ch.bound")
    /** The previous semiperiod boundary: the DP lower bound after a break. */
    val last: CounterView = new CounterView(vm, previous, builder, "ch.last")
    val margin: CounterView = new CounterView(vm, previous, builder, "ch.margin")
    /** The unary 2h countdown of the two-semiperiod continuation. */
    val cycle: CounterView = new CounterView(vm, previous, builder, "ch.cycle")

    var mode: Mode = previous.fold(Mode.Idle)(p => Mode.fromLabel(vm.label(p)("ch.mode")))
    var direction: Int = previous.fold(1)(p => vm.label(p)("ch.dir").asInt)
    var phase: Int = previous.fold(0)(p => vm.label(p)("ch.phase").asInt)
    var periodOnly: Boolean = previous.fold(false)(p => vm.label(p)("ch.only").asBool)

    /** Begin copying the semiperiod at the current center (after a DP `found`). */
    def start(): Unit = {
      period.reset()
      val symbol = center.read().getOrElse(throw new IllegalStateException("chain start requires a real center place"))
      period.write(FRONT_MARK + symbol)
      walker.copyFrom(center)
      verifier.copyFrom(center)
      h.reset()
      distance.reset()
      boundary.reset()
      last.reset()
      cycle.reset()
      periodOnly = false
      alias(lag, radius)
      alias(margin, radius)
      mode = Mode.Copy
      direction = 1
      phase = 0
    }

    /** The predicted next symbol, once the verifier has caught up in `watch`. */
    def prediction(): Option[String] = {
      if (mode != Mode.Watch || lag.sign != 0) {
        None
      } else {
        Some(period.read().takeRight(1))
      }
    }

    /** Whether four verified semiperiods permit a chain shift now. */
    def canShift: Boolean = {
      val ready = mode == Mode.Watch && lag.sign == 0 && phase == 4
      ready && (if (periodOnly) { cycleEnd } else { margin.sign >= 0 })
    }

    /** The continuation countdown holds exactly one cell. */
    def cycleEnd: Boolean = cycle.sign > 0 && cycle.pos.belowOfTop.isEmpty

    /** Check the two-semiperiod continuation invariant against real input. */
    def checkPair(left: Option[String]): Unit = {
      if (periodOnly && mode == Mode.Watch && lag.sign == 0) {
        if (cycle.sign <= 0) {
          throw new AssertionError("chain continuation exhausted without dispatch")
        }
        val same = left == prediction()
        if (same == cycleEnd) {
          throw new AssertionError("left head contradicts the two-semiperiod continuation")
        }
      }
    }

    /** Enter the two-semiperiod continuation after a shift decision. */
    def beginShift(): Unit = {
      periodOnly = true
      cycle.reset()
    }

    /** Python `_consume`: verify one place against the period tape; `false` breaks the chain. */
    private def consume(): Boolean = {
      verifier.right()
      val token = period.read()
      if (!verifier.read().contains(token.takeRight(1))) {
        mode = Mode.Broken
        false
      } else {
        distance.inc()
        if (token.startsWith(FRONT_MARK) || token.startsWith(TAIL_MARK)) {
          alias(last, boundary)
          alias(boundary, distance)
          phase = math.min(4, phase + 1)
          direction = if (token.startsWith(FRONT_MARK)) { 1 } else { -1 }
        }
        period.move(direction)
        true
      }
    }

    /** One new place has joined the matched interval (also on a chain shift). */
    def matched(): Unit = {
      margin.inc()
      if (periodOnly) {
        cycle.dec()
      }
      if (mode == Mode.Watch && lag.sign == 0) {
        consume()
      } else {
        lag.inc()
      }
    }

    /** The caller moved its center by one real place. */
    def shiftOne(): Unit = {
      for (counter <- Seq(distance, boundary, last, margin)) {
        counter.dec()
      }
      cycle.inc()
      cycle.inc()
    }

    /** One bounded local step, reading the DP unary answer tape `answer`. */
    def step(answer: TapeView): Unit = {
      mode match {
        case Mode.Copy => stepCopy(answer)
        case Mode.Back => stepBack()
        case Mode.Watch if lag.sign > 0 =>
          if (consume()) {
            lag.dec()
          }
        case Mode.Idle | Mode.Watch | Mode.Broken => ()
      }
    }

    /** Consume the unary DP answer while copying C-1, C-2, ... onto the period tape. */
    private def stepCopy(answer: TapeView): Unit = {
      if (answer.read() == "1") {
        answer.move(-1)
        h.inc()
        for (_ <- 0 until 4) {
          margin.dec()
        }
        walker.left()
        val symbol = walker.read().getOrElse(throw new IllegalArgumentException("DP answer crosses the input origin"))
        period.move(1)
        period.write(symbol)
      } else if (answer.read() == LEFT && h.sign > 0) {
        period.write(TAIL_MARK + period.read())
        mode = Mode.Back
      } else {
        throw new IllegalArgumentException("positive unary DP answer required")
      }
    }

    /** Return the period tape to its front mark, then watch. */
    private def stepBack(): Unit = {
      if (period.read().startsWith(FRONT_MARK)) {
        period.move(1)
        mode = Mode.Watch
      } else {
        period.move(-1)
      }
    }

    def finish(): Unit = {
      period.finish()
      for (counter <- Seq(h, lag, distance, boundary, last, margin, cycle)) {
        counter.finish()
      }
      builder.label("ch.mode") = mode.label
      builder.label("ch.dir") = direction
      builder.label("ch.phase") = phase
      builder.label("ch.only") = periodOnly
    }
  }
}
